src/Monotonic.cpp

```diff
@@ -30,26 +30,172 @@ using std::string;
 
 namespace {
 
-class MonotonicVisitor : public IRVisitor {
+bool is_constant(const ConstantInterval &a) {
+    return a.is_single_point(0);
+}
+
+bool is_monotonic_increasing(const ConstantInterval &a) {
+    return a.has_lower_bound() && a.min >= 0;
+}
+
+bool is_monotonic_decreasing(const ConstantInterval &a) {
+    return a.has_upper_bound() && a.max <= 0;
+}
+
+ConstantInterval to_interval(Monotonic m) {
+    switch (m) {
+    case Monotonic::Constant:
+        return ConstantInterval::single_point(0);
+    case Monotonic::Increasing:
+        return ConstantInterval::bounded_below(0);
+    case Monotonic::Decreasing:
+        return ConstantInterval::bounded_above(0);
+    case Monotonic::Unknown:
+        return ConstantInterval();
+    }
+    return ConstantInterval();
+}
+
+Monotonic to_monotonic(const ConstantInterval &x) {
+    if (is_constant(x)) {
+        return Monotonic::Constant;
+    } else if (is_monotonic_increasing(x)) {
+        return Monotonic::Increasing;
+    } else if (is_monotonic_decreasing(x)) {
+        return Monotonic::Decreasing;
+    } else {
+        return Monotonic::Unknown;
+    }
+}
+
+ConstantInterval unify(const ConstantInterval &a, const ConstantInterval &b) {
+    return ConstantInterval::make_union(a, b);
+}
+
+ConstantInterval unify(const ConstantInterval &a, int64_t b) {
+    ConstantInterval result;
+    result.include(b);
+    return result;
+}
+
+// Helpers for doing arithmetic on ConstantIntervals that avoid generating
+// expressions of pos_inf/neg_inf.
+ConstantInterval add(const ConstantInterval &a, const ConstantInterval &b) {
```

Comment:
@rootjalex could you check these helpers to see if you think they're correct?

Reply:
At first glance, I see a lot of issues if you're using empty intervals with these, because they have defined bounds. i.e. `negate(ConstantInterval::nothing())` -> `[-1, 0]`. I'd like to formally verify some of these, but the problems with empty intervals should probably be fixed first.